lib/valueflow.cpp

```diff
@@ -7087,9 +7086,15 @@ static void valueFlowContainerSize(TokenList *tokenlist, SymbolDatabase* symbold
         values.back().valueType = ValueFlow::Value::ValueType::CONTAINER_SIZE;
         if (known)
             values.back().setKnown();
-        if (Token::simpleMatch(var->nameToken()->next(), "{")) {
+        if (Token::simpleMatch(var->nameToken()->next(), "{")) { // TODO: distinguish uniform initialization from init lists
             const Token* initList = var->nameToken()->next();
-            values = getInitListSize(initList, var->valueType()->container, known);
+            const std::vector<const Token*> args = getArguments(initList);
+
+            // call to constructor taking two iterators (not an init list)
+            if (args.size() == 2 && args[0]->variable() && args[0]->variable()->isStlType() && endsWith(args[0]->variable()->getTypeName(), "iterator", 8))
```

Comment:
This check should happen in `getInitSize` and then return an empty vector if its passed a pair of iterators.

Reply:
A pair of iterators can be passed when constructing with `std::vector<int> v{l.begin(), l.end()}` and when assigning as well `v = {l.begin(), l.end()}` so it would be best to have the logic in `getInitSize` so its consistent everywhere.

Reply:
Ok, I thought that `getInitSize()` had to always return a valid size, I'll continue working on this in a week or so.

> You can use `astIsIterator` to check if the token is an iterator or not. We do something similar when doing lifetime analysis:
> 
> https://github.com/danmar/cppcheck/blob/main/lib/valueflow.cpp#L3774

I tried `astIsIterator` first, but it did't work for my test case (returned false). Maybe it's only intended for `auto` variables? `isIterator()` also does a two-stage check.

Reply:
> Maybe it's only intended for auto variables?

It should work for any variable or expression. If thats not the case, then `SymbolDatabase::setValueType` should probably be fixed.

> isIterator() also does a two-stage check.

That only checks variables. So it wont work for iterators passed as expressions like: `std::vector<int> v{l.begin(), l.end()}`.

Reply:
> but it did't work for my test case (returned false).

Actually, looking at the test case, I think this is because there is no library configuration for `std::wsregex_token_iterator`. Usually, iterators are defined from their container type(via `<container>` tag), but currently there is no support for iterator adaptors like this. I think we need to add a `<iterator>` tag to define iterator adaptors.

Reply:
> I'll continue working on this in a week or so.

Actually, `getInitSize` already checks for pairs of iterators here:

https://github.com/danmar/cppcheck/blob/main/lib/valueflow.cpp#L7056

So this mainly needs an update to library configs and `SymbolDatabase::setValueType`.

Reply:
Interesting. If only there was a comment to explain what's going on...

Reply:
Yea, sorry about that.

Reply:
As long as we don't have a configuration for iterator adaptors, could cppcheck be prevented from assuming that it deals with an initializer list with two arguments? Do a type check maybe?

Reply:
As a hack, you could an if statement [here](https://github.com/danmar/cppcheck/blob/main/lib/symboldatabase.cpp#L6247) to match the type name, with something like: 

```cpp
} else if (Token::Match(type, "cregex_iterator|wcregex_iterator|sregex_iterator|wsregex_iterator|regex_iterator|cregex_token_iterator|wcregex_token_iterator|sregex_token_iterator|wsregex_token_iterator|regex_token_iterator")) {
    valuetype->type = ValueType::Type::ITERATOR;
    // Adjust type token
}
```

Or maybe just check that the token ends with `iterator`:

```cpp
} else if (endsWith(tok->str(), "iterator") {
    valuetype->type = ValueType::Type::ITERATOR;
    // Adjust type token
}
```

We just need to make sure we dont set `ITERATOR` for types like `cregex_token_iterator::value_type` or `regex_token_iterator<T>::value_type`.

Reply:
This might fix [6624](https://trac.cppcheck.net/ticket/6624) as well.

Reply:
I was thinking more along the lines of "only assume init list if the types match the container's value_type". But I'm not sure that I could implement either solution.
Here is a similar issue involving pointers: https://trac.cppcheck.net/ticket/10513

Reply:
> I was thinking more along the lines of "only assume init list if the types match the container's value_type". But I'm not sure that I could implement either solution.

The valueType has `containerTypeToken` to get a token to the containers type. Then to get its valueType you would need to call `ValueType::parseDecl(vt->containerTypeToken, settings)`. Thats what we do [here](https://github.com/danmar/cppcheck/blob/main/lib/valueflow.cpp#L3267). You will need to check if valueType's are compatible, which would be similar to [`isLifetimeBorrowed`](https://github.com/danmar/cppcheck/blob/main/lib/valueflow.cpp#L3148).

Maybe it might be simpler to assume an iterator pair when `valueType` is missing or the type is `UNKNOWN_TYPE`. Although, we will probably need to do both in the future to support containers of iterators or pointers.

> Here is a similar issue involving pointers: https://trac.cppcheck.net/ticket/10513

If we use some extra heuristics for this, then we should add a function like `bool isIteratorPair(std::vector<const Token *> args)` so this can be reused for lifetime constructors.

Reply:
> Actually, `getInitSize` already checks for pairs of iterators here:
> 
> https://github.com/danmar/cppcheck/blob/main/lib/valueflow.cpp#L7056
> 
This check is never hit for strings. So it check must be repeated within the `if (...StringLike)` block.

Also, I think the checks below are incorrect if a custom allocator is passed, or a constructor with a `pos` parameter is used:
https://github.com/danmar/cppcheck/blob/main/lib/valueflow.cpp#L7058
Cf. https://en.cppreference.com/w/cpp/string/basic_string/basic_string